tests/src/test/scala/it/unibo/scafi/test/functional/stdlib/TestGradient.scala

```diff
@@ -0,0 +1,240 @@
+        exec(new TestProgram {
+          override def main(): Double = classicGradient(sense("source"))
+        }, ntimes = fewRounds)(stdNet)
+
+        assertNetworkValues((0 to 8).zip(List(
+          1.41, 1.0, 1.41,
+          1.0, 0.0, 1.0,
+          1.41, 1.0, infinity
+        )).toMap, Some((d1:Double, d2:Double) => d1===d2 +- 0.01))(stdNet)
+      }
+    }
+  }
+  describe("Hop Gradient"){
+    describe("On Standard Network"){
+      it("Should be possible to build a gradient of distances on node 0") {
+        stdNet.chgSensorValue("source", Set(0), true)
+        exec(new TestProgram {
+          override def main(): Double = hopGradient(sense("source"))
+        }, ntimes = fewRounds)(stdNet)
+
+        assertNetworkValues((0 to 8).zip(List(
+          0.0, 1.0, 2.0,
+          1.0, 1.0, 2.0,
+          2.0, 2.0, infinity
+        )).toMap)(stdNet)
+      }
+      it("Should be possible to build a gradient of distances on node 4") {
+        stdNet.chgSensorValue("source", Set(4), true)
+        exec(new TestProgram {
+          override def main(): Double = hopGradient(sense("source"))
+        }, ntimes = fewRounds)(stdNet)
+
+        assertNetworkValues((0 to 8).zip(List(
+          1.0, 1.0, 1.0,
+          1.0, 0.0, 1.0,
+          1.0, 1.0, infinity
+        )).toMap)(stdNet)
+      }
+      it("Should return a constant field if no source is selected") {
+        exec(new TestProgram {
+          override def main(): Double = hopGradient(sense("source"))
+        }, ntimes = fewRounds)(stdNet)
+
+        assertNetworkValues((0 to 8).zip(List(
+          infinity, infinity, infinity,
+          infinity, infinity, infinity,
+          infinity, infinity, infinity
+        )).toMap)(stdNet)
+      }
+      it("Should build a gradient for each source, node 0 and 4"){
+        stdNet.chgSensorValue("source", Set(0, 4), true)
+        exec(new TestProgram {
+          override def main(): Double = hopGradient(sense("source"))
+        }, ntimes = someRounds)(stdNet)
+
+        assertNetworkValues((0 to 8).zip(List(
+          0.0, 1.0, 1.0,
+          1.0, 0.0, 1.0,
+          1.0, 1.0, infinity
+        )).toMap)(stdNet)
+      }
+      it("Should be able to react to a change of source"){
+        stdNet.chgSensorValue("source", Set(0), true)
+        exec(new TestProgram {
+          override def main(): Double = hopGradient(sense("source"))
+        }, ntimes = fewRounds)(stdNet)
+
+        stdNet.chgSensorValue("source", Set(0), false)
+        stdNet.chgSensorValue("source", Set(4), true)
+        exec(new TestProgram {
+          override def main(): Double = hopGradient(sense("source"))
+        }, ntimes = fewRounds)(stdNet)
+
+        assertNetworkValues((0 to 8).zip(List(
+          1.0, 1.0, 1.0,
+          1.0, 0.0, 1.0,
+          1.0, 1.0, infinity
+        )).toMap, Some((d1:Double, d2:Double) => d1===d2 +- 0.01))(stdNet)
+      }
+    }
+  }
+
+  describe("BIS Gradient - refactor") {
```

Comment:
Ah, I see only now that you tried to refactor the tests, very well.
But you still improve your code through additional removal of repetition by observing that you just need to pass `bisGradient`, `classicGradient`, `flexGradient(_, metric=nbrRange, epsilon=0.05)` ... to method `testBasicBehaviour`

Split tests for a single gradients and tests comparing two gradients.

Also notice that when we compare gradients, we may want to express assertions on certain features (such as reactivity, or that a certain gradient mitigates the "rising value problem" whereas the classic gradient does not -- but to do so we also need to prepare a network exhibiting such problem!).

Reply:
for now the tests are aimed only at verifying the correctness, ie they produce the same results as the classical algorithm.
Other scenarios can certainly be added